mava/systems/ppo/sebulba/rec_mappo.py

```diff
@@ -0,0 +1,864 @@
+
+
+def learner_setup(
+    key: chex.PRNGKey, config: DictConfig, learner_devices: List
+) -> Tuple[
+    SebulbaLearnerFn[RNNLearnerState, RNNPPOTransition],
+    Tuple[RecActorApply, RecCriticApply],
+    RNNLearnerState,
+    Sharding,
+]:
+    """Initialise learner_fn, network and learner state."""
+
+    # Create temporory envoirnments.
+    env = environments.make_gym_env(config, config.arch.num_envs, add_global_state=True)
+
+    # Get number of agents and actions.
+    action_space = env.single_action_space
+    config.system.num_agents = len(action_space)
+    config.system.num_actions = int(action_space[0].n)
+
+    devices = mesh_utils.create_device_mesh((len(learner_devices),), devices=learner_devices)
+    mesh = Mesh(devices, axis_names=("learner_devices",))
+    model_spec = PartitionSpec()
+    data_spec = PartitionSpec("learner_devices")
+    learner_sharding = NamedSharding(mesh, model_spec)
+
+    # PRNG keys.
+    key, actor_key, critic_key = jax.random.split(key, 3)
+
+    # Define network and optimisers.
+    actor_pre_torso = hydra.utils.instantiate(config.network.actor_network.pre_torso)
+    actor_post_torso = hydra.utils.instantiate(config.network.actor_network.post_torso)
+
+    action_head, _ = get_action_head(action_space)
+    actor_action_head = hydra.utils.instantiate(action_head, action_dim=config.system.num_actions)
+
+    critic_pre_torso = hydra.utils.instantiate(config.network.critic_network.pre_torso)
+    critic_post_torso = hydra.utils.instantiate(config.network.critic_network.post_torso)
+
+    actor_network = Actor(
+        pre_torso=actor_pre_torso,
+        post_torso=actor_post_torso,
+        action_head=actor_action_head,
+        hidden_state_dim=config.network.hidden_state_dim,
+    )
+    critic_network = Critic(
+        pre_torso=critic_pre_torso,
+        post_torso=critic_post_torso,
+        hidden_state_dim=config.network.hidden_state_dim,
+        centralised_critic=True,
+    )
+
+    actor_lr = make_learning_rate(config.system.actor_lr, config)
+    critic_lr = make_learning_rate(config.system.critic_lr, config)
+
+    actor_optim = optax.chain(
+        optax.clip_by_global_norm(config.system.max_grad_norm),
+        optax.adam(actor_lr, eps=1e-5),
+    )
+    critic_optim = optax.chain(
+        optax.clip_by_global_norm(config.system.max_grad_norm),
+        optax.adam(critic_lr, eps=1e-5),
+    )
+
+    # Initialise observation: Select only obs for a single agent.
+    init_obs = env.reset().observation
+    init_agents_view = init_obs.agents_view[0][jnp.newaxis, jnp.newaxis, :]
+    init_action_mask = init_obs.action_mask[0][jnp.newaxis, jnp.newaxis, :]
+    init_global_state = init_obs.global_state[0][jnp.newaxis, jnp.newaxis, :]  # type: ignore
+    single_obs = ObservationGlobalState(init_agents_view, init_action_mask, init_global_state)
+    init_done = jnp.zeros((1, config.arch.num_envs, config.system.num_agents), dtype=bool)
```

Comment:
We can't do it like this, were we require an `env.reset` There must be a way to get the global obs from the spec, or we need to make one